doc/language_reference/language_reference.md

```diff
@@ -380,45 +390,106 @@ exponent ::= (* empty *)
+### String literals
 
-```EBNF
-string_literal   ::= '"' utf8_character* '"'
-                     | "[|" utf8_character* "|]"
-```
+We support two types of UTF-8 string literals: quoted strings with
+escaping, and raw strings.
 
-Multiple string literals are automatically concatenated, e.g.,
-`"foo" [|bar|]` is equivalent to `"foobar"`.
+In quoted strings, e.g.,`"foo\nbar"` C-like escape are recognized; for
+example single quotes and backslash characters can be escaped using
+backslashes; `\n` is newline, and `\t` is tab.  Unicode character with
+code 100 can be written as `\u{100}`.
 
-Interpolated strings are string literals, that contain
-expressions inside curly brackets preceeded by a dollar sign (`${}`), whose values are substituted at runtime.
-Quoted strings are interpolate by default, e.g.,
-`"x: ${x}, y: ${y}, f(x): ${f(x)}"` is equivalent to
-`"x: " ++ x ++ ", y: " ++ y ++ ", f(x): " ++ f(x)`.
-
-Raw interpolated strings must be preceed by a dollar sign:
+Interpolated strings are string literals, that contain expressions
+inside curly brackets preceeded by a dollar sign (`${}`), whose values
+are substituted at runtime.  Quoted strings are interpolated by
+default, e.g., `"x: ${x}, y: ${y}, f(x): ${f(x)}"` is equivalent to
+the following string expressoin: `"x: " ++ x ++ ", y: " ++ y ++ ",
+f(x): " ++ f(x)`.
 
 ```EBNF
+string_literal   ::= '"' utf8_character* '"'
+                     | raw_string
+                     | raw_interpolated_string
+
+raw_string       ::= "[|" utf8_character* "|]"
 raw_interpolated_string ::= ("$[|" utf8_character* "|]")+
 ```
 
+Raw strings are delimited by `[|` and `|]` (where the closing sequence
+cannot appear nested inside the string).  In raw strings all
+characters, including backslash and line breaks are interpreted as-is.
+Raw strings do not perform interpolation.
+
+Raw interpolated strings, which are preceeded by a dollar sign, will
+perform interpolation: `$[|x: ${x}, y: ${y}, f(x): ${f(x)}|]` produces
+the same string as above.
+
+Compare the value of the raw string `[|a = ${2+3}|]`, which is the
+string containing the following characters: `a = ${2+3}`, with the
+value of the raw interpolated string `$[|a = ${2+3}|]`, which is the
+string containing `a = 5`.
+
 Expressions in curly brackets can be arbitrarily complex, as long as
 they produce results of type `string`, e.g.:
 `$[|foo{var x = "bar"; x}|]` will evaluate to "foobar" at runtime.
 
-Other terms:
+Consecutive string literals are automatically concatenated, e.g.,
+`"foo" [|bar|]` is equivalent to `"foobar"`.
+
+### Automatic string conversion
+
+Values of arbitrary types that occur inside interpolated strings or as
+a second argument to the string concatenation operator (`++`) are
+automatically converted to strings.
+Values of primitive types (`string`, `bigint`, `bit`, and `bool`) are converted using
+builtin methods.
+
+For user-defined types, conversion is performed by calling a user-defined function
```

Comment:
This is no longer true. All string conversion functions are called `to_string()`. I forgot to update the docs.

Reply:
Fixed this one; also combed the tree for remaining uses of the old name.